File: generate/reference/duality-of-compilation/ocaml/factorial_accumulator.ml
let rec factorial a i =
  if i = 0 then
    a
  else
    factorial ((i * a) mod 1000000007) (i - 1)

let main n =
  factorial 1 n

let () =
  match Sys.argv with
  | [| _ |] -> print_endline (string_of_int (main 10))
  | [| _; s |] -> (
      match int_of_string_opt s with
      | Some n -> print_endline (string_of_int (main n))
      | None -> print_endline "argument is not a number")
  | _ -> print_endline "arguments are too many"

File: generate/reference/duality-of-compilation/rust/factorial_accumulator.rs
use std::env;

fn factorial(i: isize) -> isize {
    let mut acc = 1;
    for j in 1..(i + 1) {
        acc = (j * acc) % 1000000007
    }
    acc
}

fn run(n: isize) -> isize {
    factorial(n)
}

fn main() {
    let args: Vec<String> = env::args().collect();

    let arg = match args.get(1) {
        None => 10,
        Some(val) => match val.parse::<isize>() {
            Ok(val) => val,
            Err(_) => {
                println!("argument is not a number");
                return;
            }
        },
    };

    println!("{}", run(arg));
}
